### PR TITLE
Tracking PR: Revision getters

### PR DESCRIPTION
**Pull request by du5t**
Starting PR a teeny bit early for discussion again.

The two functions I added (`getRevisions()` and `getLatestRevision()`) do what they say on the label and have unit tests for them, but need an ssb instance passed to them so they can fetch related messages and then filter down to revisions.

I need to get `getLatestRevision()` into `flattenThread()` so that thread rendering with post edits will "just work", but I don't want to break it by suddenly requiring it to be async. Thoughts?

### CODE REVIEW

**Comment by pfrazee** [comment]
Conventionally we call incomplete PRs "Tracking PRs," so I've renamed this one

File: index.js [hunk 1]
```diff
@@ -386,6 +391,81 @@ exports.getLastThreadPost = function (thread) {
...
+
+exports.getRevisions = function(ssb, thread, callback) {
+  function collectRevisions(thread) {
+    // this function walks the revisions of a given thread, collecting them up
+    // asyncly
+
+    var threadRevisions = thread.related
+        .map(function(relatedMsg) {
+      
+        if (relatedMsg.value.content.type === 'post-edit' &&
```
Comment (pfrazee) [inline index.js]: in situations like this, I'll usually assign content to a variable so I dont have to have giant references like that. Eg:

``` js
var c = relatedMsg.value.content
if (c.type == 'post-edit' && c.revision && ...)
```

File: index.js [hunk 2]
```diff
@@ -386,6 +391,81 @@ exports.getLastThreadPost = function (thread) {
...
+            // ^ make sure it's an edit
+            relatedMsg.value.content.revision &&
+            // ^ may be unnecessary if schema validates
+            (relatedMsg.value.content.revision === thread.key ||
+             relatedMsg.value.content.root     === thread.key)
```
Comment (pfrazee) [inline index.js]: wait, what does `root` point to? The topmost post in a post-thread, or the very first version of a post?

File: index.js [hunk 3]
```diff
@@ -386,6 +391,81 @@ exports.getLastThreadPost = function (thread) {
...
+            // ^ either the msg revises the root, or it's a subsequent revision
+           ) {
+          return(relatedMsg);
+        }
+      })
+      .filter(function(msg) { return msg || false}) // eliminate falsehood
```
Comment (pfrazee) [inline index.js]: Neat trick you can do for this: `.filter(Boolean)`. Same effect, filters out any falsey items.

File: index.js [hunk 4]
```diff
@@ -386,6 +391,81 @@ exports.getLastThreadPost = function (thread) {
...
+      .sort(function (oneRev, otherRev) { // remove duplicates by sorting and
+                                          // reducing over the sorted arr
+        return oneRev.key < otherRev.key
+      });
+    if (threadRevisions.length > 0) {
+      threadRevisions =
+        threadRevisions.reduce(function(prevRevs, thisRev, thisInd) {
```
Comment (pfrazee) [inline index.js]: is there a better way to do this? I appreciate why you took this approach, but it's not much cleaner than how flattenThread removes duplicates (via a Set), and I think it's less clear

File: index.js [hunk 5]
```diff
@@ -386,6 +391,81 @@ exports.getLastThreadPost = function (thread) {
...
+  if (!thread.hasOwnProperty('related')) { // if the thread doesn't have its related objects,
+                         // fetch them
+    ssb.relatedMessages(thread, function(err, enrichedThread) {
+      if (err) throw err
+      else if (!enrichedThread.hasOwnProperty('related')) callback([]) // if still no related objects
+      else callback(collectRevisions(enrichedThread))
+    })
+  } else { // note: this branch is technically synchronous and the above is not
+           // :(
+    callback(collectRevisions(thread))
```
Comment (pfrazee) [inline index.js]: Use the node `(err, result)` callback convention. Except in rare cases, we try to use that in all situations.

File: index.js [hunk 6]
```diff
@@ -386,6 +391,81 @@ exports.getLastThreadPost = function (thread) {
...
+        })
+    }
+    return threadRevisions
+  }
+            
+
+  if (!thread.hasOwnProperty('related')) { // if the thread doesn't have its related objects,
+                         // fetch them
+    ssb.relatedMessages(thread, function(err, enrichedThread) {
+      if (err) throw err
```
Comment (pfrazee) [inline index.js]: Don't throw, pass it into the callback, `return callback(err)`

**Comment by pfrazee** [comment]
Going well, good work! I've added some comments that need addressing.

**Comment by du5t** [comment]
thanks! that was wonderful code review! fixing...

File: index.js (hunk 2 above)
Comment (du5t) [inline index.js]: as I understood it, related messages always include the root of their thread. So each revision is a "revision thread" and its root is the first version of a post.

Comment (pfrazee) [inline index.js]: `.relatedMessages()` takes a message ID and gets the full tree of messages that link to it. The concept of a thread is somewhat abstract, but usually we're referring to a "post thread," as in the tree of the discussion. So, in a `type: post`, the `root` link goes to the first post in the discussion.

> So each revision is a "revision thread"

You mean, each post can have its own "revision thread?"

File: index.js (hunk 4 above)
Comment (du5t) [inline index.js]: oh no wai, I didn't know es6 was available everywhere. yeah, totally agree

**Comment by du5t** [comment]
This one is pretty ready too. Works with https://github.com/ssbc/patchwork-threads/pull/14 and https://github.com/ssbc/patchwork/pull/270.

**Comment by du5t** [comment]
whoops! needs one more test case for multiple edits on the same post.

**Comment by dominictarr** [comment]
@du5t can you explain to me how this represents changes? I'm having a little trouble seeing how it actually works. A PR should always include an explanation. code does not speak for itself.

**Comment by du5t** [comment]
whoops, sorry! here is a summary of changes:

## exportable functions added
- `reviseFlatThread(ssb, thread, cb)`: replaces each message in a flattened thread with its latest revision
- `getRevisions(ssb, thread, cb)`: gets the messages published as a revision to an existing message
- `getLatestRevision(ssb, msg, cb)`: gets the latest revision (by sequence) of some message

## other changes
- `flattenThread()` now checks against an array `thingsThatArePosts` which contains `['post', 'post-edit']` when checking a post for eligibility to be inserted into a flat thread

## convenience functions (not exported)
- `isaRevisionTo(a, b)`: works just like `isaMentionTo(a,b)` but for revisions
- `removeThreadDuplicates(threadArr)`: DRYs up the process of taking an array of msgs, putting their keys into a Set, and returning only the msgs whose keys are in the Set

## tests

a ton of tests on a range of threads from simple to intricate, that were terribly necessary for debugging. here are their description lines:
- 'getRevisions returns an array of revisions back to root'
- 'getRevisions returns an array'
- 'getRevisions returns an array with the right number and type of revisions'
- 'getLatestRevision returns the latest rev of a msg'
- 'getLatestRevision returns the original msg if no revisions'
- 'getLatestRevision returns latest rev of a msg even if edited many times'
- 'getLatestRevision is not confused by multiple edits different from root'
- 'reviseFlatThread returns the original message if only one is present'
- 'reviseFlatThread returns the latest revision of every member of a thread'
- 'reviseFlatThread returns properly even if root is revised'

also a couple of edge cases that weren't obvious to me in earlier tests:
- 'edge 1: root edited multiple times out of sequence with rest of thread'
- 'edge 2: reply edited out of sequence with rest of thread'

**Comment by dominictarr** [comment]
@du5t my main concern is how does it represent the changes. what does the (raw) edited post look like. what points to what etc, etc? what happens to the database?

**Comment by du5t** [comment]
@dominictarr  i put some notes on that in https://github.com/ssbc/ssb-msg-schemas/pull/14 where i added schemas of post-edits

File: index.js [hunk 7]
```diff
@@ -109,7 +132,8 @@ exports.flattenThread = function (thread) {
     addedIds.add(msg.key)
   }
   function flattenAndReorderReplies (msg) {
-    if (msg.value.content.type == 'post' && isaReplyTo(msg, thread)) {
+    if (thingsThatArePosts.indexOf(msg.value.content.type !== -1) &&
```
Comment (pfrazee) [inline index.js]: Bug here!

File: index.js [hunk 8]
```diff
@@ -130,7 +154,7 @@ exports.flattenThread = function (thread) {
       if (addedIds.has(msg.key))
         return // skip duplicates
       // insert if a mention to its parent
-      if (msg.value.content.type == 'post' && isaMentionTo(msg, parent))
+      if (thingsThatArePosts.indexOf(msg.value.content.type !== -1) && isaMentionTo(msg, parent))
```
Comment (pfrazee) [inline index.js]: And here.

File: index.js [hunk 9]
```diff
@@ -386,6 +410,122 @@ exports.getLastThreadPost = function (thread) {
...
 }
 
+
+/* post revision utils */
+
+exports.getRevisions = function(ssb, thread, callback) {
+  function collectRevisions(thread, callback) {
+    // this function walks the revisions of a given thread, collecting them up
+    // asyncly
+    var deepThread = pluckRecursively(thread, 'related')[0]
```
Comment (pfrazee) [inline index.js]: Why use `pluckRecursively` if you're just going to take `[0]` ? Shouldnt this just be

``` js
var deepThread = thread.related
```

File: index.js [hunk 10]
```diff
@@ -386,6 +410,122 @@ exports.getLastThreadPost = function (thread) {
...
+
+    var threadRevisionLogCB = multicb({pluck: 1})
+    deepThread
+    .filter(function(relatedMsg) { return isaRevisionTo(relatedMsg, thread) })
+    .forEach(function(edit) {
+      exports.createRevisionLog(ssb, edit, threadRevisionLogCB())
```
Comment (pfrazee) [inline index.js]: Is this necessary? `relatedMessages` does a recursive thread-collection, but then you're throwing that away and doing your own recursive fetch-from-db here.

File: index.js [hunk 11]
```diff
@@ -386,6 +410,122 @@ exports.getLastThreadPost = function (thread) {
...
+          // passing tests
+          return revLog[revLog.length-1].key === thread.key
+        })
+
+      // the longest such log is the one that contains all of the revisions
+      const logLengths = connectedLogs.map(function(log) { return log.length })
+      const thisLog = connectedLogs.find(function(log) {
+                        return log.length === Math.max.apply(Math, logLengths)
+                      })
+      callback(null, thisLog)
```
Comment (pfrazee) [inline index.js]: Can you explain why this search is occuring? You should be able to get the full revision log off of the thread root, by using `ssb.links()` or `ssb.relatedMessages()`, because all of the revisions have a `root` link to their top. You can then trust the order of those revisions by their sequence numbers, because they were all published by the same user.

File: index.js [hunk 12]
```diff
@@ -393,4 +533,64 @@ function isaReplyTo (a, b) {
...
+
+  return isEdit && editsThis && (sameAuthor || isWiki) && isNewer
+}
+
+function traverseRelatedEdits(msg, callback) {
+  var latestRev = msg
+  if (latestRev.hasOwnProperty('related')) {
+    latestRev.related.forEach(function (relatedMsg) {
+      
+//      if (latestRev.value.content.text === 'edge3-b-revised2') { debugger }
```
Comment (pfrazee) [inline index.js]: leftover debug

File: index.js [hunk 13]
```diff
@@ -393,4 +533,64 @@ function isaReplyTo (a, b) {
...
+}
+
+function isNewerRevision(msg, rev) {
+  // more explicit check of revision relationship than above
+  var root = mlib.link(rev.root)
+  const relMsg       = (rev.value.content)
+  const isEdit       = (relMsg.type === 'post-edit')
+  const editsThis    = (relMsg.revision === msg.key)
+  const sameAuthor   = (rev.value.author === msg.value.author)
+  const isWiki       = (rev.value.author === 'wiki')
```
Comment (pfrazee) [inline index.js]: what is this?

File: index.js [hunk 14]
```diff
@@ -393,4 +533,64 @@ function isaReplyTo (a, b) {
...
+  const isNewer      = (rev.value.timestamp > msg.value.timestamp)
```
Comment (pfrazee) [inline index.js]: Use `rev.value.sequence`, not `rev.value.timestamp`. Timestamp is the wallclock and may not increase monotonically

File: index.js [hunk 15]
```diff
@@ -386,6 +410,122 @@ exports.getLastThreadPost = function (thread) {
...
+      if (err) callback (err)
+      // if still no related objects
+      else if (!richMsg.hasOwnProperty('related')) { callback(null, richMsg) }
+      else { traverseRelatedEdits(richMsg, callback) }
+    })
+  } else if (!msg.hasOwnProperty('related')) {
+    callback(null, msg)
+  } else {
+    traverseRelatedEdits(msg, callback)
+  }
```
Comment (pfrazee) [inline index.js]: Similar to other functions, this shouldnt need to be recursive. You can just look through `.related` for the post-edit with the highest sequence.

File: index.js [hunk 16]
```diff
@@ -386,6 +410,122 @@ exports.getLastThreadPost = function (thread) {
...
+}
+
+exports.createRevisionLog = function (ssb, msg, callback) {
```
Comment (pfrazee) [inline index.js]: I dont think this method is needed in getRevisions, and Im not sure it's needed at all.

This method is basically a copy of `fetchThreadRootID`, which mainly exists to deal with threads that are malformed (a message with a bad/missing `root`). I dont think we need that kind of tolerance in the revisions, so I think this function isnt needed.

File: test/revisions.js [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+     function(t) {
+     t.plan(3)
+
+     var db = sublevel(level('test-patchwork-threads-revision-array-count', {
+       valueEncoding: defaults.codec
+     }))
+     var ssb = SSB(db, defaults)
+
+     
+       var alice = ssb.createFeed(ssbKeys.generate())
```
Comment (pfrazee) [inline test/revisions.js:87]: Indentation is fubar here

**Comment by pfrazee** [comment]
@du5t Time to get this guy merged! I left some feedback here, and I'm going to do some more feedback on the ssb-msg-schemas PR. If you can knock those out, we'll be ready.

(Can you also fix some of the indentation? Especially in the tests, there's some really fubar tabbing)

**Comment by du5t** [comment]
yeah, the tests are straight callback pyramid city :(

i'll go thru and de-cruft the code first, and see what tests fail.

File: index.js (hunk 7 above)
Comment (du5t) [inline index.js]: maybe i'm misunderstanding what this function is doing--should it be considering only "posts that are replies" or "things that are posts that are also replies"? i was thinking that post-edits should be treated like posts in these two cases, since they're identical except that they 'replace' a post.

File: index.js (hunk 13 above)
Comment (du5t) [inline index.js]: a future feature :P

as long as the schema matches, anyone can publish edits to anyone else's posts--the only thing keeping anyone from doing so is the sort of UI affordances around it, like this one, determining which revisions to actually consider when choosing a msg's latest revision.

it's no stretch to see that community wiki a la stack overflow is an easy thing to add this way (though an additional property like `value.isWiki` might be better?), so i threw it in just to see.

File: index.js (hunk 8 above)
Comment (du5t) [inline index.js]: wait, how is this one a bug? if a `post` is 'replaced' with a `post-edit`, don't we want it to be weave-able the same way?

File: index.js (hunk 9 above)
Comment (du5t) [inline index.js]: yup, tests pass without the extra! i wrote that function when i was still losing my mind over how the thread contents would come through. fixing...

File: index.js (hunk 11 above)
Comment (du5t) [inline index.js]: i think i might be able to get it from the output of L424. i don't want to make everything depend on the revisions all being published by the same user though, since there are use cases like annotations, group work, and wikis.

File: index.js (hunk 16 above)
Comment (du5t) [inline index.js]: the current implementation certainly needs a look, but for a 'revision history view', something under the same name would be nice to have. i'll see what i can do with `revisionRoot` and `revisionBranch`.

File: index.js (hunk 8 above)
Comment (clehner) [inline index.js]: Misplaced paren?
`if (thingsThatArePosts.indexOf(msg.value.content.type) !== -1 && isaMentionTo(msg, parent))`

File: index.js (hunk 13 above)
Comment (pfrazee) [inline index.js]: Yeah, I appreciate the idea but let's remove it. There's a lot of complexity around multiple-user editing in EC systems, which we cant handle yet, and wouldnt handle this way.

File: index.js (hunk 7 above)
Comment (pfrazee) [inline index.js]: The problem isnt what that condition _intends_ to do, it's just that the conditional is mis-written. It should be:

``` js
if (thingsThatArePosts.indexOf(msg.value.content.type) !== -1 && ...
```

The equality check against `-1` occurs outside of the parameters for `indexOf`

File: index.js (hunk 11 above)
Comment (pfrazee) [inline index.js]: We shouldnt plan for multi-user editing yet.

File: index.js (hunk 16 above)
Comment (pfrazee) [inline index.js]: Doesnt `getRevisions` already satisfy that need? I think `createRevisionLog` is redundant

File: index.js (hunk 7 above)
Comment (du5t) [inline index.js]: oh god, sorry, this is exactly what [Array.prototype.includes()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) is for...

File: index.js (hunk 15 above)
Comment (du5t) [inline index.js]: i tested it again and i remember why this was such a pain. if you try to get the related edits from a message that isn't far enough down the tree, it only tells you about the ones before it. i don't remember how i thought it worked anymore, but to see this, try running the tests with a version of this function that uses the method you suggest. a couple of the edge tests written for this very case will fail.

``` js
exports.getLatestRevision = function (ssb, thread, callback) {
  const sortRevsGetLast = function(richMsg) {
    const sortedRevs = richMsg.related
            .filter(function (rev) {
              return rev.value.content.type === 'post-edit'
            })
            .sort(function(msgA, msgB) {
              return msgA.value.sequence < msgB.value.sequence
            })                         
    return sortedRevs[0]
  }
  var msg = thread
  if (!msg.hasOwnProperty('related')) {
    ssb.relatedMessages({id: msg.key, count: true}, function(err, richMsg) {
      if (err) callback (err)
      // if still no related objects
      else if (!richMsg.hasOwnProperty('related')) { callback(null, richMsg) }
      else {
        callback(null, sortRevsGetLast(richMsg))
      }
      // else { traverseRelatedEdits(richMsg, callback) }
    })
  } else if (!msg.hasOwnProperty('related')) {
    callback(null, msg)
  } else {
//    traverseRelatedEdits(msg, callback)
    callback(null, sortRevsGetLast(msg))
  }
}
```

File: index.js (hunk 16 above)
Comment (du5t) [inline index.js]: i haven't split up the schema yet, but `threadlib.createRevisionLog(ssb, revisionA2, function(err, threadLog) {` on L34 of the revisions test will fail if you replace it with `threadlib.getRevisions(ssb, revisionA2, function(err, threadLog) {` (same signature). same kind of issues i was having in https://github.com/ssbc/patchwork-threads/pull/4#discussion_r56118643.

File: index.js (hunk 15 above)
Comment (pfrazee) [inline index.js]: Yeah, this is harder than it should be, because the old linking structure. Once `revisionRoot` is in place, the way this will work is by getting the most recent `.related` that points to the target message as a `revisionRoot`. This will be the new filter:

``` js
.filter(function (rev) {
  return rev.value.content.type === 'post-edit'
    && mlib.relationsTo(rev, thread).indexOf('revisionRoot') !== -1)
})
```

If it's cool with you, I can fork these PRs and make the updates I'm describing. We're almost there, and I don't want to keep tasking you with updates when the turn-around is so long.

Comment (du5t) [inline index.js]: i can give it my full extended attention within 18 hours. does that fit your sprint? i was waiting on consensus in https://github.com/ssbc/ssb-msg-schemas/pull/14, but i like the decision you went with, so...

**Comment by du5t** [comment]
done! matches new schema. i've cleaned up un-needed wacky functions also.

**Comment by pfrazee** [comment]
Great! Taking a look now

File: index.js [hunk 18]
```diff
@@ -58,6 +63,28 @@ exports.getPostThread = function (ssb, mid, opts, cb) {
...
+  })
+  
+  callbackAggregator(function (err, results) {
+    if (err) {
+      callback(err)
+    } else {
+      callback(null, results)
+    }
+  })
+}
```
Comment (pfrazee) [inline index.js:84]: Will this method lose any non-posts in the flattened thread?

File: index.js [hunk 19]
```diff
@@ -2,6 +2,11 @@ var mlib = require('ssb-msgs')
 var pull = require('pull-stream')
 var multicb = require('multicb')
 
+// shim for array.includes from es6
+Array.prototype.includes = function(item) {
+  return this.indexOf(item) !== -1
+}
```
Comment (pfrazee) [inline index.js]: Hate to be picky about this, but it's not a good idea to modify global prototypes within a library. If it's not a perfect 1-1 match to es6's implementation, it could create bugs. I suggest just creating:

``` js
function includes (array, item) {
...
```

File: README.md [hunk 20]
```diff
@@ -48,4 +51,13 @@ threadlib.decryptThread (ssb, thread, cb)
...
\ No newline at end of file
+
+// get messages that were published as revision to an existing message
+threadlib.getRevisions(ssb, thread, callback)
+
+// get the latest revision of an existing message
+threadlib.getLatestRevision(ssb, msg, callback)
+
+// produces the previous revisions of a msg going back to the original msg
+threadlib.createRevisionLog(ssb, msg, cb)
```
Comment (pfrazee) [inline README.md]: this can go!

**Comment by pfrazee** [comment]
A few small updates and then this is ready to go

File: index.js (hunk 18 above)
Comment (du5t) [inline index.js:84]: i think it does so deliberately. the idea is that anything that is a post has a latest revision (case 1), and anything that is a post-edit is either the latest revision of some post (case 2), or it isn't (case 3).
- case 1: replace with latest rev
- case 2: will be handled by case 1
- case 3: would not be displayed anyway

File: index.js (hunk 19 above)
Comment (du5t) [inline index.js]: yeah, i don't want to make anyone responsible for that. refactoring...

File: index.js (hunk 18 above)
Comment (du5t) [inline index.js:84]: oh wait, i see what you mean. there needs to be a passthru so other messages make it into the callback.

Comment (pfrazee) [inline index.js:84]: Yeah, correct

Comment (du5t) [inline index.js:84]: added in 113b7fb

**Comment by pfrazee** [comment]
I think this is ready to go. @dominictarr if you want to review again, now's the time. Otherwise, I'll merge tomorrow.

**Comment by pfrazee** [comment]
Good work @du5t, thanks again for being patient with me!

**Comment by du5t** [comment]
thanks for being patient with _me!_ looking forward to a design discussion on multi-user/wiki edits

**Comment by dominictarr** [comment]
lets merge this. Sorry I didn't realize I left this this long!
Great work @du5t and @pfraze !

**Comment by pfrazee** [comment]
published `2.1.0`
